Add support for table-valued parameters

brandonfoong opened the pull request:
## TVP support in ODBC

To bind a TVP and its columns as the argument of a procedure call in ODBC, we perform the following steps:

1. Call [`SQLBindParameter`](https://docs.microsoft.com/en-us/sql/odbc/reference/syntax/sqlbindparameter-function) with `ParameterType=SQL_SS_TABLE` to specify that we are creating a TVP
2. Call [`SQLSetStmtAttr`](https://docs.microsoft.com/en-us/sql/odbc/reference/syntax/sqlsetstmtattr-function) with `Attribute=SQL_SOPT_SS_PARAM_FOCUS` and `ValuePtr=TableIndex` to begin modifying the columns of the table
3. Call `SQLBindParameter` to bind the columns of the table
4. Call `SQLSetStmtAttr` to reset the focus back to the parameter list instead of the table
5. Call `SQLExecDirect` to submit the query

Some sample uses of the ODBC API can be found [here](https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-how-to/use-table-valued-parameters-odbc) and [here](https://www.easysoft.com/developer/languages/c/examples/TVPInsertRecs.html).

## FreeTDS implementation

Internally, FreeTDS represents a SQL statement as a [`_hstmt`](https://www.freetds.org/reference/a01031.html) struct, which contains the query string and metadata required to execute the query. The query parameters are stored in the `apd` and `ipd` fields, and are added by `_SQLBindParameter` [here](https://github.com/brandonfoong/freetds/blob/61e292c238b1eb737252dd47a8a802e9323f6c18/src/odbc/odbc.c#L1457). When a remote procedure call is executed by `SQLExecDirect`, it will convert the parameters from the ODBC to TDS format (via [`odbc_sql2tds`](https://github.com/brandonfoong/freetds/blob/61e292c238b1eb737252dd47a8a802e9323f6c18/src/odbc/sql2tds.c#L164)) and then the TDS lib will send the query to the server.

## Changes made

I used a similar approach to store and convert the columns of a TVP. When we bind a column, we store the arguments in a `SQLTVPCOLUMN` struct. When we convert the TVP in `odbc_sql2tds`, we generate an `apd` and `ipd` value for each row, and then use `odbc_sql2tds` recursively to convert that row into TDS format. From there, we can use the existing serialization functions to send each element of the table via the TDS protocool.

<!--
## Changes made

### New additions

1. `SQLTVP`/`SQLTVPCOLUMN` structs - stores table and column information in ODBC format
2. `TDS_TVP`/`TDS_TVP_ROW` structs - stores table and row information in TDS format

### Changes to existing functions

1. `SQLBindParameter` - If we specify `ParameterType=SQL_SS_TABLE`, allocate a `SQLTVP` buffer to store table-related information ([here](https://github.com/brandonfoong/freetds/blob/579124d1da0c39c7ee7ffb85e3f83d2e07dc22e1/src/odbc/odbc.c#L1543))
2. `SQLSetStmtAttr` - If we specify `Attribute=SQL_SOPT_SS_PARAM_FOCUS`, update the `param_focus` variable in `stmt` as per the `ValuePtr` argument ([here](https://github.com/brandonfoong/freetds/blob/579124d1da0c39c7ee7ffb85e3f83d2e07dc22e1/src/odbc/odbc.c#L6825) - `ui` is the value of `ValuePtr` converted back to an integer)
3. `SQLBindParameter` - If the current `param_focus` is set to a table, store the incoming values in the `SQLTVP` buffer ([here](https://github.com/brandonfoong/freetds/blob/579124d1da0c39c7ee7ffb85e3f83d2e07dc22e1/src/odbc/odbc.c#L1606))
4. `odbc_sql2tds` - If the parameter is a table, recursively convert its columns into the TDS format via `odbc_convert_table`, then convert the table from ODBC to TDS format ([here](https://github.com/brandonfoong/freetds/blob/579124d1da0c39c7ee7ffb85e3f83d2e07dc22e1/src/odbc/sql2tds.c#L622) and [here](https://github.com/brandonfoong/freetds/blob/579124d1da0c39c7ee7ffb85e3f83d2e07dc22e1/src/odbc/sql2tds.c#L153))
-->

brandonfoong commented on include/freetds/odbc.h:
```diff
@@ -568,6 +569,26 @@ TDS_DBC *desc_get_dbc(TDS_DESC *desc);
 /*
  * odbc.c
  */
+/* Temporarily store the bound TVP columns */
+/* Follow the naming in _SQLBindParameter for convenience */
+typedef struct {
+	SQLSMALLINT fParamType; // InputOutputParam
+	SQLSMALLINT fCType;     // C data type
+	SQLSMALLINT fSqlType;   // SQL data type
+	SQLULEN cbColDef;       // columnsize
+	SQLSMALLINT ibScale;    // precision
+	SQLPOINTER rgbValue;    // content buffer
+	SQLLEN cbValueMax;      // buffer length
+	SQLLEN * pcbValue;      // content length buffer
+} SQLTVPCOLUMN;
+
+typedef struct {
+	char * schema;
+	char * typename;
+	int num_cols;
+	SQLTVPCOLUMN ** cols;
+} SQLTVP;
+
```
`SQLTVP`/`SQLTVPCOLUMN` structs - stores table and column information in ODBC format

brandonfoong commented on include/freetds/tds.h:
```diff
@@ -1324,6 +1324,21 @@ TDSICONV *tds_iconv_from_collate(TDSCONNECTION * conn, TDS_UCHAR collate[5]);
 
 
 /* mem.c */
+typedef struct tds_tvp_row
+{
+	TDSPARAMINFO * params;
+	struct tds_table_row * next;
+} TDS_TVP_ROW;
+
+typedef struct tds_tvp
+{
+	char * schema;
+	char * typename;
+	int num_cols;
+	TDS_TVP_ROW * metadata;
+	TDS_TVP_ROW * row;
+} TDS_TVP;
+
```
`TDS_TVP`/`TDS_TVP_ROW` structs - stores table and row information in TDS format

brandonfoong commented on src/odbc/odbc.c:
```diff
@@ -1398,14 +1398,74 @@ SQLGetEnvAttr(SQLHENV henv, SQLINTEGER Attribute, SQLPOINTER Value, SQLINTEGER B
 			exit; \
 	} while(0)
 
+static SQLRETURN
```
These functions handle the case where we add new columns to a TVP

jiangyue12392 replied:
Should this be in `mem.c`?

jiangyue12392 replied:
Do you really need to copy and allocate for all columns every time?

brandonfoong replied:
Will check if this can be defined in `mem.c` instead. However, from what I remember, `odbc.h` includes `tds.h` and not the other way around, so the TDS library doesn't know how the `SQLTVP` struct is defined.

As for the column allocation, referring to the the `OrdEntry_TVP` function code [here](https://docs.microsoft.com/en-us/sql/relational-databases/native-client-odbc-how-to/use-table-valued-parameters-odbc) (specifically the `SQLBindParameter` calls under `// Col 1 - ProdCode` and `// Col 2 - Qty`), it seems like the number of columns is not known initially, but added one at a time, so we need to copy and resize the underlying array.

brandonfoong commented on src/odbc/odbc.c:
```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
-		ODBC_EXIT_(stmt);
-	}
-	drec = &ipd->records[ipar - 1];
+		if (fSqlType == SQL_SS_TABLE) {
```
If we specify `ParameterType=SQL_SS_TABLE`, allocate a `SQLTVP` buffer to store table-related information. The type  of the TVP is passed as a `char*` buffer in `rgbValue`, and should have the format `schema.typename`, or just `typename`.

The table buffer is stored in `stmt->apd->records[index - 1].sql_desc_data_ptr`, where `stmt->apd` stores information about outgoing data.

brandonfoong commented on src/odbc/odbc.c:
```diff
@@ -6709,6 +6822,9 @@ _SQLSetStmtAttr(SQLHSTMT hstmt, SQLINTEGER Attribute, SQLPOINTER ValuePtr, SQLIN
 			break;
 		}
 		break;
+	case SQL_SOPT_SS_PARAM_FOCUS:
+		stmt->param_focus = ui;
+		break;
```
If we specify `Attribute=SQL_SOPT_SS_PARAM_FOCUS`, update the `param_focus` variable in `stmt` as per the `ValuePtr` argument. `ui` is the value of `ValuePtr` converted back to an integer.

brandonfoong commented on src/odbc/odbc.c:
```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
+				/* TODO: report error */
+			}
+
+			tvp = odbc_alloc_table(schema, typename);
+			if ((drec->sql_desc_data_ptr = (char *) tvp) == NULL) {
+				odbc_errs_add(&stmt->errs, "HY001", NULL);
+				ODBC_EXIT_(stmt);
+			}
+		}
+
+		/* field IPD related fields */
+		ipd = stmt->ipd;
+		orig_ipd_size = ipd->header.sql_desc_count;
+		if (ipar > ipd->header.sql_desc_count && desc_alloc_records(ipd, ipar) != SQL_SUCCESS) {
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY001", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		drec = &ipd->records[ipar - 1];
+
+		drec->sql_desc_parameter_type = fParamType;
+		if (odbc_set_concise_sql_type(fSqlType, drec, 0) != SQL_SUCCESS) {
+			desc_alloc_records(ipd, orig_ipd_size);
+			desc_alloc_records(apd, orig_apd_size);
+			odbc_errs_add(&stmt->errs, "HY004", NULL);
+			ODBC_EXIT_(stmt);
+		}
+		if (is_numeric) {
+			drec->sql_desc_precision = cbColDef;
+			drec->sql_desc_scale = ibScale;
+		} else {
+			drec->sql_desc_length = cbColDef;
+		}
 	} else {
-		drec->sql_desc_length = cbColDef;
+		tvp = (SQLTVP *) stmt->apd[stmt->param_focus - 1].records->sql_desc_data_ptr;
+		if (add_table_column(tvp, ipar, fParamType, fCType, fSqlType, cbColDef, ibScale, rgbValue, cbValueMax, pcbValue) != SQL_SUCCESS) {
+			odbc_errs_add(&stmt->errs, "HY001", NULL);
+			ODBC_EXIT_(stmt);
+		}
```
If the current `param_focus` is set to a table, store the incoming values in the `SQLTVP` buffer instead of the statement.

brandonfoong commented on src/tds/data.c:
```diff
@@ -1476,6 +1476,101 @@ tds_sybbigtime_put(TDSSOCKET *tds, TDSCOLUMN *col, int bcp7)
+}
+
+TDSRET
+tds_mstabletype_put(TDSSOCKET *tds, TDSCOLUMN *col, int bcp7)
```
These functions handle serialization of the TVP according to the TDS protocool.

brandonfoong commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,129 @@ _odbc_blob_free(TDSCOLUMN *col)
 }
 
+static TDS_INT
+odbc_convert_table(TDS_STMT * stmt, SQLTVP * src, TDS_TVP * dest, SQLLEN num_rows)
```
Handle table conversion from ODBC to TDS format. We first allocate a dummy row to store the metadata of the table, then, for each row, we allocate an `apd` and `ipd` buffer (similar to `SQLBindParameter`), and then use `odbc_sql2tds` to convert the row to TDS format.

jiangyue12392 commented on include/freetds/tds.h:
```diff
@@ -1376,7 +1391,6 @@ TDSBCPINFO *tds_alloc_bcpinfo(void);
 void tds_free_bcpinfo(TDSBCPINFO *bcpinfo);
 void tds_deinit_bcpinfo(TDSBCPINFO *bcpinfo);
 
-
```
Seems to be an accidental change here

jiangyue12392 commented on src/odbc/odbc.c:
```diff
@@ -1398,14 +1398,74 @@ SQLGetEnvAttr(SQLHENV henv, SQLINTEGER Attribute, SQLPOINTER Value, SQLINTEGER B
+}
+
+static SQLTVP *
+odbc_alloc_table(char schema[], char typename[])
```
Consider putting this in mem.c, too

jiangyue12392 commented on src/odbc/odbc.c:
```diff
@@ -1398,14 +1398,74 @@ SQLGetEnvAttr(SQLHENV henv, SQLINTEGER Attribute, SQLPOINTER Value, SQLINTEGER B
-	int is_numeric = 0;
+	int is_numeric = 0, tvp_name_len;
+	SQLTVP * tvp;
+	char * tvp_name, * schema, * typename, * pch;
```
Maybe should rename `typename` as `type_name`, `typename` is a keyword in c++.

jiangyue12392 commented on src/odbc/odbc.c:
```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
+				ODBC_EXIT_(stmt);
+			}
+
+			strncpy(tvp_name, rgbValue, tvp_name_len);
```
Probably just `memcpy` the `tvp_name_len + 1` size here.

brandonfoong replied:
Out of curiosity, are `strncpy` and `memcpy` compiled differently?

For this, I'm adding in the null terminator because the char buffer may not be null-terminated, and the length is specified via `cbValueMax` (line 1550).

jiangyue12392 commented on src/odbc/odbc.c:
```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
+
+			/* Tokenize and extract the schema & typename from the TVP's full name */
+			pch = strchr(tvp_name, '.');
+			if (pch == NULL) {
+				schema = "";
+				typename = tvp_name;
+			} else {
+				*pch = 0;
+				schema = tvp_name;
+				typename = ++pch;
+			}
+
+			/* Ensure that the typename does not contain '.' */
+			pch = strchr(typename, '.');
+			if (pch != NULL) {
+				/* TODO: report error */
+			}
+
+			tvp = odbc_alloc_table(schema, typename);
```
How about allocate tvp first then use the `tvp->schema` and `tvp->typename` directly?

jiangyue12392 commented on src/odbc/odbc.c:
```diff
@@ -1453,54 +1513,105 @@ _SQLBindParameter(SQLHSTMT hstmt, SQLUSMALLINT ipar, SQLSMALLINT fParamType, SQL
+		}
 	}
 
+
```
Extra line

jiangyue12392 commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,129 @@ _odbc_blob_free(TDSCOLUMN *col)
+	}
+	dest->metadata->params = params;
+
+	for (i = 0, prow = &dest->row; i < num_rows; prow = &(*prow)->next, i++) {
```
Probably don't need `i` and just do something like
`for (prow = &dest->row; num_rows; prow = &(*prow)->next, num_rows--)`

brandonfoong replied:
Turns out that `i` is being used to calculate the buffer offset for the i-th element, since all buffers are passed as byte pointers

jiangyue12392 commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,129 @@ _odbc_blob_free(TDSCOLUMN *col)
+
+	dest->metadata->next = NULL;
+	params = NULL;
+	for (j = 0; j < src->num_cols; j++) {
+		new_params = tds_alloc_param_result(params);
+
+		if (j+1 > apd->header.sql_desc_count && desc_alloc_records(apd, j+1) != SQL_SUCCESS)
+			return -1;
+		drec = &apd->records[j];
+
+		if (odbc_set_concise_c_type(src->cols[j]->fCType, drec, 0) != SQL_SUCCESS)
+			return -1;
+
+		if (drec->sql_desc_type == SQL_C_CHAR || drec->sql_desc_type == SQL_C_WCHAR || drec->sql_desc_type == SQL_C_BINARY)
+			drec->sql_desc_octet_length = src->cols[j]->cbValueMax;
+		drec->sql_desc_indicator_ptr = (SQLLEN *) SQL_NULL_DATA;
+		drec->sql_desc_octet_length_ptr = (SQLLEN *) SQL_NULL_DATA;
+		drec->sql_desc_data_ptr = NULL;
+
+		/* field IPD related fields */
+		ipd = stmt->ipd;
+		if (j+1 > ipd->header.sql_desc_count && desc_alloc_records(ipd, j+1) != SQL_SUCCESS)
+			return -1;
+		drec = &ipd->records[j];
+
+		drec->sql_desc_parameter_type = src->cols[j]->fParamType;
+		if (odbc_set_concise_sql_type(src->cols[j]->fSqlType, drec, 0) != SQL_SUCCESS)
+			return -1;
+		if (src->cols[j]->fSqlType == SQL_DECIMAL || src->cols[j]->fSqlType == SQL_NUMERIC) {
+			drec->sql_desc_precision = src->cols[j]->cbColDef;
+			drec->sql_desc_scale = src->cols[j]->ibScale;
+		} else {
+			drec->sql_desc_length = src->cols[j]->cbColDef;
+		}
+
+		odbc_sql2tds(stmt, &ipd->records[j], &apd->records[j], new_params->columns[j], false, apd, 0);
+		params = new_params;
+	}
```
Maybe can refactor some logic out with the below part

jiangyue12392 commented on src/tds/unittests/allcolumns.c:
```diff
@@ -132,6 +132,7 @@ void tds_all_types(TDSSOCKET *tds, tds_any_type_t *func)
 		case SYBUNITEXT:
 		case SYBVARIANT: /* TODO */
 		case SYBSINT1: /* TODO */
+		case SYBTABLETYPE: /* TODO */
```
What does this `TODO` mark mean?

brandonfoong replied:
This is just to ignore the unit test for converting to/from table values so that I can proceed on to the other tests. Regardless, converting any other type to/from a table type is invalid, so this shouldn't affect the functionality too much.

freddy77 committed:
Take into account bulk information from server for Sybase

Issuing "insert into bulk" statement the server returns some
information for the table.
Use them, they are surely more correct than the one we assume
and compute.

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>

jiangyue12392 commented on include/freetds/tds.h:
```diff
@@ -1324,6 +1324,21 @@ TDSICONV *tds_iconv_from_collate(TDSCONNECTION * conn, TDS_UCHAR collate[5]);
 
 
 /* mem.c */
+typedef struct tds_tvp_row
+{
+	TDSPARAMINFO * params;
+	struct tds_tvp_row * next;
+} TDS_TVP_ROW;
+
+typedef struct tds_tvp
+{
+	char * schema;
+	char * typename;
+	int num_cols;
+	TDS_TVP_ROW * metadata;
+	TDS_TVP_ROW * row;
+} TDS_TVP;
```
Seems to be more appropriate to put it under `/* data.c */`

jiangyue12392 commented on src/tds/mem.c:
```diff
@@ -1895,5 +1896,4 @@ tds_realloc(void **pp, size_t new_size)
 
 	return p;
 }
-
```
Unnecessary modification

jiangyue12392 commented on src/odbc/unittests/tvp.c:
```diff
@@ -0,0 +1,188 @@
+    odbc_disconnect();
+
+	return 0;
+}
```
Seems that their new file convention is to have extra ending new line. Probably good to follow that.

jiangyue12392 replied:
Also, is it that you have a mixture of spaces and tabs in this file? Please follow their convention and use all tabs

jiangyue12392 commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,140 @@ _odbc_blob_free(TDSCOLUMN *col)
         TDS_ZERO_FREE(col->column_data);
 }
 
+/* Copied from _SQLBindParameter - is there a better way of converting the column metadata? */
```
Probably remove this when you eventually PR to freetds

jiangyue12392 commented on include/freetds/tds.h:
```diff
@@ -1324,6 +1324,21 @@ TDSICONV *tds_iconv_from_collate(TDSCONNECTION * conn, TDS_UCHAR collate[5]);
+typedef struct tds_tvp
+{
+	char * schema;
+	char * typename;
```
Seems that you only changed some of the `typename`, please do a clean up for it.

jiangyue12392 commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,140 @@ _odbc_blob_free(TDSCOLUMN *col)
+		if (!(new_params = tds_alloc_param_result(params)))
+			return -1;
+
+		odbc_convert_table_row(apd, ipd, j+1, cols[j]->fParamType, cols[j]->fCType,
```
`j + 1`

jiangyue12392 commented on src/odbc/sql2tds.c:
```diff
@@ -149,6 +149,140 @@ _odbc_blob_free(TDSCOLUMN *col)
+				return -1;
+
+			rgbValue = ((BYTE *) cols[j]->rgbValue) + i * cols[j]->cbValueMax;
+			odbc_convert_table_row(apd, ipd, j+1, cols[j]->fParamType, cols[j]->fCType,
```
`j + 1`

jiangyue12392 commented on src/tds/data.c:
```diff
@@ -1530,6 +1623,16 @@ tds_sybbigtime_check(const TDSCOLUMN *col)
 	return 1;
 }
 
+/* TODO */
+
+int
+tds_mstabletype_check(const TDSCOLUMN *col)
+{
+	return 0;
+}
+
+/* END TODO */
+
```
Are the todo markers here still relevant?

freddy77 committed:
tds: Fix for change in OpenSSL 1.1.1l

Some strings in OpenSSL 1.1.1l are not NUL terminated as in 1.1.1k.
This was causing a wrong check for hostname causing failure connecting
to database.
Terminate always extracted hostname string to avoid this issue.

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>

freddy77 committed:
Fix static compile with readline library

In static compiling order is more important, library depending
on others should came first, not after.

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>

Commit:
odbc: add unit tests for connection string parsing

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>
Signed-off-by: Szilárd Parrag <szilard.parrag@oneidentity.com>
Signed-off-by: Kokan <kokaipeter@gmail.com>
Signed-off-by: Attila Szakacs <attila.szakacs@oneidentity.com>

freddy77 committed:
odbc: Add support for quoted string in connection string

Signed-off-by: Frediano Ziglio <freddy77@gmail.com>

brandonfoong committed:
odbc: add handling of table-valued params in ODBC and dblib

brandonfoong committed:
odbc: add unit tests for table-valued params
